File: sdk/dart/lib/adapter/chat_session.dart
// SPDX-License-Identifier: Apache-2.0
// Copyright (c) 2026 RunAnywhere AI, Inc.

import 'dart:async';

import 'llm_session.dart';
import 'types.dart';

enum ChatRole { system, user, assistant, tool }

class ChatMessage {
  final ChatRole role;
  final String content;
  const ChatMessage(this.role, this.content);

  factory ChatMessage.system(String c)    => ChatMessage(ChatRole.system, c);
  factory ChatMessage.user(String c)      => ChatMessage(ChatRole.user, c);
  factory ChatMessage.assistant(String c) => ChatMessage(ChatRole.assistant, c);
  factory ChatMessage.tool(String c)      => ChatMessage(ChatRole.tool, c);
}

/// Chat wrapper over LLMSession. Message history + token stream.
class ChatSession {
  final LLMSession llm;
  bool _systemInjected = false;

  ChatSession(String modelId, String modelPath,
              {String systemPrompt = '',
               ModelFormat format = ModelFormat.gguf})
      : llm = LLMSession(modelId, modelPath, format: format) {
    if (systemPrompt.isNotEmpty) {
      final rc = llm.injectSystemPrompt(systemPrompt);
      _systemInjected = rc == 0;
    }
  }

  Stream<String> generate(List<ChatMessage> messages) async* {
    final rendered = renderMessages(messages, skipSystem: _systemInjected);
    final tokens = _systemInjected
        ? llm.generateFromContext(rendered)
        : llm.generate(rendered);
    await for (final t in tokens) {
      if (t.kind == LLMTokenKind.answer) yield t.text;
    }
  }

  Future<String> generateText(List<ChatMessage> messages) async {
    final buf = StringBuffer();
    await for (final chunk in generate(messages)) { buf.write(chunk); }
    return buf.toString();
  }

  int cancel() => llm.cancel();
  int resetHistory() {
    _systemInjected = false;
    return llm.clearContext();
  }
  void close() => llm.close();

  static String renderMessages(List<ChatMessage> messages,
                                 {bool skipSystem = false}) {
    final sb = StringBuffer();
    for (final m in messages) {
      if (skipSystem && m.role == ChatRole.system) continue;
      sb.write('<|im_start|>${m.role.name}\n${m.content}<|im_end|>\n');
    }
    sb.write('<|im_start|>assistant\n');
    return sb.toString();
  }
}

File: sdk/ts/src/adapter/ChatSession.ts
// SPDX-License-Identifier: Apache-2.0
// Copyright (c) 2026 RunAnywhere AI, Inc.

import { ModelFormat, LLMTokenKind } from './Types.js';
import { LLMSession } from './LLMSession.js';

export type ChatRole = 'system' | 'user' | 'assistant' | 'tool';

export interface ChatMessage {
  role: ChatRole;
  content: string;
}

export const ChatMessage = {
  system:    (content: string): ChatMessage => ({ role: 'system',    content }),
  user:      (content: string): ChatMessage => ({ role: 'user',      content }),
  assistant: (content: string): ChatMessage => ({ role: 'assistant', content }),
  tool:      (content: string): ChatMessage => ({ role: 'tool',      content }),
};

/**
 * Chat-style wrapper over LLMSession. Manages message history and exposes
 * `generate(messages)` → `AsyncIterable<string>` (token text).
 */
export class ChatSession {
  private readonly llm: LLMSession;
  private systemPromptInjected = false;

  constructor(public readonly modelId: string,
              public readonly modelPath: string,
              systemPrompt = '',
              format: ModelFormat = ModelFormat.GGUF) {
    this.llm = new LLMSession(modelId, modelPath, format);
    if (systemPrompt) {
      const rc = this.llm.injectSystemPrompt(systemPrompt);
      this.systemPromptInjected = rc === 0;
    }
  }

  async *generate(messages: ChatMessage[]): AsyncIterable<string> {
    const rendered = ChatSession.renderMessages(messages, this.systemPromptInjected);
    const source = this.systemPromptInjected
      ? this.llm.generateFromContext(rendered)
      : this.llm.generate(rendered);
    for await (const token of source) {
      if (token.kind === LLMTokenKind.Answer) yield token.text;
    }
  }

  async generateText(messages: ChatMessage[]): Promise<string> {
    let buf = '';
    for await (const chunk of this.generate(messages)) buf += chunk;
    return buf;
  }

  cancel(): number { return this.llm.cancel(); }
  resetHistory(): number {
    this.systemPromptInjected = false;
    return this.llm.clearContext();
  }
  close(): void { this.llm.close(); }

  static renderMessages(messages: ChatMessage[], skipSystem: boolean): string {
    let out = '';
    for (const m of messages) {
      if (skipSystem && m.role === 'system') continue;
      out += `<|im_start|>${m.role}\n${m.content}<|im_end|>\n`;
    }
    out += '<|im_start|>assistant\n';
    return out;
  }
}
